### ISSUE TITLE
Change learning rate during training

### ISSUE THREAD

**Issue by r-zemblys** [ibab/tensorflow-wavenet#77]
Would it be helpful to add ability to manually change learning rate during training? I have naive implementation of this feature [here](https://github.com/r-zemblys/tensorflow-wavenet/tree/dev).
![lr2](https://cloud.githubusercontent.com/assets/8914323/18787185/694e945e-81aa-11e6-9a0f-b47b12486e4b.png)
![lr1](https://cloud.githubusercontent.com/assets/8914323/18787184/694bd110-81aa-11e6-97c9-3665a5351b0f.png)

**Comment by lemonzi** [comment]
The standard approach is to add 2 parameters: a number of iterations during which the learning rate is kept constant, and a multiplier that is applied after that for exponential decay. This forces the network to adjust to the closest local minima when training is running out of iterations rather than keep exploring the space.

The best way of changing it manually is stopping the training and restoring it from a checkpoint with a different learning rate.

I think TensorFlow has a built-in implementation of this exponential decay, or at least plenty of examples in the published models; it would be indeed very helpful to have it.

**Comment by jyegerlehner** [comment]
@r-zemblys  
Yes please. Echoing lemonzi, I'd say your particular schedule of learning rate (in the plot) looks a little odd :). I'm guessing you are showing the "square wave" learning rate schedule to show it can be adjusted at will, not that you recommend that particular schedule?

@lemonzi 
> The standard approach is to add 2 parameters: a number of iterations during which the learning rate is kept constant, and a multiplier that is applied after that for exponential decay.

Hmm.. I'm not aware of that standard. It may be a good strategy. I've seen people just using a single step-change from an initial high learning rate to an order-of-magnitude lower learning rate later. I'm not saying that's better. But I wouldn't want to hard-wire to just the "constant-followed-by-exponential-decay" approach.

>  The best way of changing it manually is stopping the training and restoring it from a checkpoint with a different learning rate.

Do you say that because you think there's something wrong with the technique in r-zembly's dev branch? I've done it the way you describe, but I like his approach of having a learning_rate placeholder and feeding it a value you can set any way you want as training progresses.

Perhaps a configurable approach would be to have a learning_rate_schedule in wavenet_params.json with a list of (iterations, learn_rate) tuples, and use that to compute learn_rate fed to the placeholder.

**Comment by lemonzi** [comment]
I recall seeing that some time ago in another framework (Kaldi?), but I was having a look and most of the official TF models seem to use step decay, which is implemented in [`tf.train.exponential_decay`](https://www.tensorflow.org/versions/r0.10/api_docs/python/train.html#exponential_decay). 

There are two main cases where you want to change the learning rate: when you are trying things out and realize the network is training too slowly (or too fast), in which case it makes sense to perform a manual stop and restore to reuse the work done (and that's what I meant before), or during training as part of a global scheduling. Global schedulers, except for the case where you just change it a couple times to a magnitude lower as you mentioned, are better defined using a formula, and inputting all the values in the JSON file doesn't seem like a good solution -- actually, we were even talking about getting rid of the dilation array and just specifying the layers per stack and the number of stacks.

I would suggest using TensorFlow's standard by default, and having the corresponding parameters in the JSON (initial LR, decay factor, steps per decay...) . To allow for weirder schedules, we can  implement it using a thin wrapper that lets people specify other schedules manually by supplying a function, or an optional flag with the list of tuples, or something like that.

Actually, since the learning rate does not affect the network itself, we could have a separate JSON file (or an nested object) with training-specific parameters, such as learning rate and regularization.

**Comment by jyegerlehner** [comment]
@lemonzi 
> We could also add a learning_rate_scheduler field to the training JSON that could be constant, step_decay, or manual. For manual, the list of tuples is supplied.

I like that idea. But we can proceed incrementally if that's easier; don't have to do all that in the first PR.

**Comment by nakosung** [comment]
Latest deepmind paper(https://arxiv.org/pdf/1610.00527v1.pdf) says they adjust lr-rate after they see learning flatlines.

*We use RMSProp for the optimization with an initial learning rate of 3 · 10−4 and multiply the
learning rate by 0.3 when learning flatlines.*

Instead of having fixed lr-policy, we can have adaptive one.

**Comment by nakosung** [comment]
Keras has 'early stopping' which monitors progression.

https://github.com/fchollet/keras/blob/master/keras/callbacks.py#L312
